fix headings and clean up troubleshooting page (#1855)
pulumi/pulumi-hugo: themes/default/content/docs/support/troubleshooting.md

preventing you from being productive, you've come to the right place.

## Contact Us

[Join Community Slack](https://slack.pulumi.com), where our whole team, in addition to a passionate
community of users, are there to help. Any and all questions are welcome!

We also encourage everyone to contribute to the [Pulumi open source
projects](https://github.com/pulumi) by [opening new
issues](https://github.com/pulumi/pulumi/issues/new) and upvoting existing issues.

Or email our support team: [support@pulumi.com](mailto:support@pulumi.com).

## Verbose Logging

Verbose logging of the internals of the Pulumi engine and resource providers can be enabled by
passing the `-v` flag to any `pulumi` CLI command. Pulumi emits logs at log levels between `1` and
`11`, with `11` being the most verbose. At log level 10 or below, Pulumi will avoid intentionally exposing any *known* credentials. At log level 11, Pulumi will intentionally expose some known credentials to aid with debugging, so these log levels should be used only when absolutely needed.

By default, logs are written to the top-level temp directory (usually `/tmp` or the value of

> `$TMPDIR`). The `--logtostderr` flag can be used to write logs to `stderr` instead.

Reviewer: ```suggestion
`$TMPDIR`). The `--logtostderr` flag can be used to write logs to stderr instead.
```

Debatable, but I would not backtick it FWIW.

Use the flag `--logflow` to apply the same log level to resource providers.

{{% notes type="warning" %}}
Enabling verbose logging may reveal sensitive information (tokens, credentials...) that is provided from

> your execution environment to your cloud provider and which Pulumi may not be aware of. Before sharing the logs, audit and redact the sensitive information.

Reviewer: ```suggestion
your execution environment directly to your cloud provider, and which Pulumi may not be aware of. Before sharing the logs, be careful to audit and redact any sensitive information.
```

## Common Problems

### 403 error fetching plugin

> This error commonly occurs when using an arm64 based processor while using an older version of a provider that does not support arm64. It's not possible to upgrade the packages as your state will still be locked to the old version of the provider.

Reviewer: ```suggestion
This error commonly occurs when using an ARM64-based processor while using an older version of a provider that does not support ARM64. It is not possible to upgrade the providers as your state file requires the version of the provider used at the time the resources were created.
```

(At least I think that's the implication...)

> The method for fiing this depends on whether you are using an Intel based computer.

Reviewer: ```suggestion
The method for fixing this issue depends on whether you are using an Intel based processor:
```

> #### Intel based computer

Reviewer: ```suggestion
#### Intel based processor
```

1. Open your Pulumi program on a non-arm64 based computer.
1. Update your packages (pip / nuget / npm / go) and run `pulumi up`.
1. Once the update is complete, open the new, updated Pulumi program on your arm64-based system.

> #### Non-Intel based computer

Reviewer: ```suggestion
#### Non-Intel based processor
```

> 1. Remove Pulumi - if you're using Homebrew, `brew remove pulumi` or `rm -rf ~/.pulumi`

Reviewer: Aren't these separate things? `brew remove pulumi` would remove the Pulumi executable (`/opt/homebrew/bin/pulumi`), but `rm -rf ~/.pulumi` would remove plugins, templates, etc.

1. Add Pulumi to path: `export PATH=$PATH:~/.pulumi/bin`
1. Update packages in your Pulumi program to latest version (for example `npm install @pulumi/aws@latest)
1. Install Pulumi provider: `arch -x86_64 pulumi plugin install resource {provider_name} v{version}` (where  {provider_name} is the name of the provider, i.e. aws and {version} is the same version number that your package has updated to). `arch` is used to run the selected architecture of a binary, in this case so that you can run the non-ARM64 version of Pulumi on your laptop.
1. [Login to Pulumi]({{< relref "/docs/intro/concepts/state/#logging-in" >}}).
1. Run a Pulumi preview: `arch -x86_64 pulumi pre`.
1. Remove Pulumi again `rm -rf ~/.pulumi`.
1. [Re-install Pulumi]({{< relref "/docs/get-started/install" >}})
1. [Login to Pulumi]({{< relref "/docs/intro/concepts/state/#logging-in" >}}).
1. Run a Pulumi preview to check everything is ok: `pulumi pre`

### 409 Conflict: Another update is currently in progress. {#conflict}

Run `pulumi cancel` to cancel the update.

{{% notes type="warning" %}}
Warning! If you cancel another person's update, their update will fail immediately.
{{% /notes %}}

> One of the services that `pulumi.com` provides is *concurrency control*. The service will allow

Reviewer 2: I wonder if instead of saying "One of the services that `pulumi.com` provides ..." we should say something like 
"One of the services that the [Pulumi Service]({{< relref "/docs/intro/pulumi-service/" >}}) provides ..."

1. Somebody else is currently updating the stack. If you are working on a stack with more than one collaborator, it could
be that your collaborators have initiated an update without your knowledge. You can confirm this by visiting the Pulumi
web console and seeing who initiated the most recent update.

> 2. You were updating the stack, but the Pulumi CLI crashed in the middle of the update.

Reviewer: ```suggestion
1. You were updating the stack, but the Pulumi CLI crashed in the middle of the update.
```

const ids = aws.ec2.getSubnetIds(..., { parent });
```

> This warning may be benign. However, if you are experiencing crashes or hangs in Pulumi (especially in Node.js version 12.11.0 and above) and you see this warning, then it is likely that this is the source.

Reviewer: ```suggestion
This warning may be benign. However, if you are experiencing crashes or hangs in Pulumi (especially in Node.js version 12.11.0 and above) and you see this warning, then it is likely that the condition described above is the root cause.
```

Author: @jkodroff we should avoid directional words, because they are not accessible. i can link to the previous section though if you'd like.

Reviewer: @susanev This is helpful information. If you have a source handy, can you send me some guidelines on how to refer to previous information that refers to e.g. the preceding paragraph?

The main thing I was trying to address was the use of "this" and "that" with no surrounding context. I _think_ they're called "unqualified articles", and I was taught that they should be avoided in technical writing, but a Google search only turns up how journalists shouldn't write about science, which is obviously not what we're talking about here.

Author: we should link to the section we are referencing. i think i put this in the style guide. but maybe im making that up.
some sources that follow this model.
- https://styleguide.mailchimp.com/writing-for-accessibility/#header-3-avoid-directional-language
- https://uxdesign.cc/ux-copy-for-accessibility-2927491cadf8 "→ Avoid using directional language.
Above, below, left, right… Far more often than you might think, these words are meaningless and misinterpreted by the users (and not just those who are blind or visually impaired). Users with screen readers won’t have right or left as a reference."
- https://developers.google.com/style/accessibility "Don't use directional language to orient the reader, such as above, below, or right-hand side. This type of language doesn't work well for accessibility or for localization reasons."

Author: ive clarified the language and removed the last reference to `that`. let me know if it makes sense.

Author: i think imma go ahead and merge tonight, happy to come back with a separate pr if i got it wrong.

of the call is then wrapped into an `Output` so it can easily be passed as a resource input and to make it
[simple to access properties]({{< relref "/docs/intro/concepts/inputs-outputs#lifting" >}}) off of it.

#### Register the provider first

> If the problem exists in a layer deeper (e.e.g, a component not under your control), use this solution.

Reviewer: ```suggestion
If the problem exists in a layer deeper (e.g, a component not under your control), use this solution.
```

const val = stackReference.getOutputSync("outputName");
```

> This warning may be benign. However, if you are experiencing crashes or hangs in Pulumi (especially in Node.js version 12.11.0 and above) and you see this warning, then it is likely that this is the source.

Reviewer: Same suggestion as above.

Author: @jkodroff i have no idea what _above_ means here? sry can you be more specific?

If the stack-reference-name truly is dynamic and cannot be known ahead of time to supply directly into the app, then this
approach will not work, and the only way to workaround the issue is to follow the steps in [Use getOutput/requireOutput](#use-getoutput).

> ### Cannot connect to the Pulumi Service

Reviewer 2: I think this section "Cannot connect to the Pulumi Service" should be moved up to follow (or maybe precede) the "Nothing happens due to network proxy" since they are both kind of networking related. It also puts the two topics related to pre-Pulumi 2.0 issues at the end.

Author: sure will do! was feeling a bit funny about this being at the bottom anyway.

can happen:

- The CLI experiences a network partition when attempting to save your stack's state.
- The CLI process is killed by your operating system while performing an update.
- The CLI crashes when performing an update.

This error means that the Pulumi engine initiated an operation and was not able to determine if the operation completed successfully. This means that resources may have been created that Pulumi does not know about.

To fix this situation, you should first cancel the last update..

```bash
$ pulumi cancel
...
The currently running update for 'interruptedstack' has been canceled!
```

If `pulumi cancel` fails with `error: [400] Bad Request: the update has already completed`, you can safely ignore
that error and continue with the next step.

You should then export and import your stack. This will clear the stacks's pending operations.

```bash
$ pulumi stack export | pulumi stack import
warning: removing pending operation 'creating' on '...' from snapshot
Import successful.
```

For every warning that this command prints out, you should verify with your cloud provider whether or not this

> operation was successful. If the operation was successful, and a resource was created, you should delete that

Reviewer: ```suggestion
operation was successful. If the operation was successful, and a resource was created, you should delete the
```

resources:

```bash
$ pulumi refresh
Refreshing stack 'interruptedstack'
Performing changes:

   Type  Name  Status     Info

   info: no changes required:
         12 resources unchanged
```

At this point your stack should be valid, up-to-date, and ready to accept future updates.

## Manually Editing Your Deployment {#editing-your-deployment}

Sometimes the only recourse for fixing a stack that is unable to do deployments is to edit the
deployment directly. We would love to hear about the issues you are experiencing
that you can't resolve, both so we can assist you in fixing your stack and also to fix the issues in Pulumi
that made it impossible for you to recover your stack in any other way.

The Pulumi engine uses both your program and your stack's existing state to make decisions about what
resources to create, read, update, or delete. The most common problem that makes it impossible to
make changes to your stack is that the stack's existing state has been corrupted in some way. There
are a variety of ways that a stack's state could be corrupted, but in almost all cases it is possible

> to manually edit the stack's existing state to fix the corruption.

Reviewer: ```suggestion
to manually edit the stack's state to fix the issue.
```

> This is an advanced operation and should be an absolute last resort. We recommend you check-in with the [Pulumi Community Slack](https://slack.pulumi.com) first before editing your snapshot.

Reviewer: ```suggestion
This is an advanced operation and should be an absolute last resort. We recommend you check-in with the [Pulumi Community Slack](https://slack.pulumi.com) first before editing your state file.
```

Author: @jkodroff i recently learned that snapshot is slightly different than state file, so would like to leave this as is. but i will be adding snapshot to the glossary soon. also, snapshot is mentioned a bunch more times in this file.

If you intend to unprotect or delete a resource, consider using the [`pulumi state`]({{< relref "/docs/reference/cli/pulumi_state" >}}) command instead of editing your state directly. `pulumi state` makes fixes to your state without
requiring you to edit the JSON representation of your stack's current state.

To get a JSON representation of your stack's current state, export your current stack
to a file:

```bash
$ pulumi stack export --file state.json
```

> This file contains a lot of information. At the top-level, this JSON object has two fields:

Reviewer: A brief search did not reveal an obvious source for detailed info on the Pulumi state file, but if there is one, we should link to it in this section.

Author: i also could not find one
